packages/hdwallet-xdefi/src/adapter.ts

```diff
@@ -0,0 +1,42 @@
+import * as core from "@shapeshiftoss/hdwallet-core";
+import { XDeFiHDWallet } from "./xdefi";
+
+export class XDeFiAdapter {
+  keyring: core.Keyring;
+
+  // wallet id to remove from the keyring when the active wallet changes
+  currentDeviceID?: string;
+
+  private constructor(keyring: core.Keyring) {
+    this.keyring = keyring;
+  }
+
+  public static useKeyring(keyring: core.Keyring) {
+    return new XDeFiAdapter(keyring);
+  }
+
+  public async initialize(): Promise<number> {
+    return Object.keys(this.keyring.wallets).length;
+  }
+
+  public async pairDevice(): Promise<core.HDWallet> {
+    const provider: any = (window as any).xfi && (window as any).xfi.ethereum;
+    if (!provider) {
+      window.open("https://xdefi.io", "_blank");
+    }
```

Comment:
It's not appropriate to fire window.open here; better to throw and let the application handle it. My instinct is also to use `globalThis` instead of `window` but idk.

Reply:
You have a very good point, I will throw an exception here and handle it on the client side. Will also check between `this` vs `window`.